### PR TITLE
initial cli testing frame

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
Initial DSL for cli testing. Not exposed through `WebTauDsl` yet.  `CliJavaTest` shows basic java syntax. Approach is similar to `Http.`. 

It is most likely going to change so considering it is a private API for now. Want to put it out there so we can start conversation around. 

One of the syntax I want to be able to achieve follows
```
cli.run('command...') {
    on('partial match') {
         http.get( ... ) { 
         }
    }

    onOptional('partial match') {
        http.del(....) {
        }
    }

    output.should contain('blah')
}
```

Alternatively we may try to use `waitTo` (not really sure how to approach this)
```
cli.run('command...') {
    line.waitTo == ~/partial match/ 
    http.get( ... ) { 
    }

    output.should contain('blah')
}
```

### CODE REVIEW

**Comment by tsiq-karold** [comment]
Can you explain what the desired behaviour of `on` and `onOptional` are?

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/Cli.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    public void run(String command, CliValidationExitCodeOutputHandler handler) {
+        cliStep(command, (runResult) -> handler.handle(exitCode(runResult), cliOutput(runResult), cliError(runResult)));
+    }
+
+    private void cliStep(String command, Consumer<ProcessRunResult> validationCode) {
+        TestStep.createAndExecuteStep(null,
+                tokenizedMessage(action("running cli command "), stringValue(command)),
+                () -> tokenizedMessage(action("ran cli clommand"), stringValue(command)),
+                () -> {
+                    ProcessRunResult runResult = ProcessUtils.run(command.split("\\s+"));
```
Comment (tsiq-karold) [inline webtau-cli/src/main/java/com/twosigma/webtau/cli/Cli.java]: Hmmm, what if someone wants to run `foo.sh "something that should be one arg and is therefore in quotes"`?

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+package com.twosigma.webtau.cli;
+
+import java.io.IOException;
+
+public class ProcessUtils {
+    private ProcessUtils() {
+    }
+
+    public static ProcessRunResult run(String[] command) {
```
Comment (tsiq-karold) [inline webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: I must have written almost identical code at least 10 times in my career :)

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        try {
+            Process process = processBuilder.start();
+
+            StreamGobbler outputGobbler = new StreamGobbler("output", process.getInputStream());
+            StreamGobbler errorGobbler = new StreamGobbler("error", process.getErrorStream());
+
+            new Thread(errorGobbler).start();
+            new Thread(outputGobbler).start();
+
+            process.waitFor();
```
Comment (tsiq-karold) [inline webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: Do we want to impose some timeouts here?

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+            return new ProcessRunResult(process.exitValue(), outputGobbler.getLines(), errorGobbler.getLines());
```
Comment (tsiq-karold) [inline webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: Could the reason for your sporadic test failures be that you're not joining the two gobbler threads?  They could in theory still be consuming the stream when you call `getLines` here.

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/Cli.java (hunk 1 above)
Comment (MykolaGolubyev) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/Cli.java]: good point, thanks

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java (hunk 2 above)
Comment (MykolaGolubyev) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: I feel you :)

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java (hunk 3 above)
Comment (MykolaGolubyev) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: Created issue earlier to track this. Want a skeleton down for now.

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java (hunk 4 above)
Comment (MykolaGolubyev) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: Trying

**Comment by MykolaGolubyev** [comment]
> Do we want to support user supplied environment variables for the process? It's not uncommon for scripts to expect some vars. People may even wish to set a PATH so they can run all the tests without FQN for the scripts they're executing.
> 
> Do we want to support multiple script executions? Right now if you do two cli.run calls, they'll have totally separate environments (or shells or whatever you want to call it).
> 
> My hunch is yes to the env vars and no to the multiple scripts but curious on your thoughts.

Yes for env vars. Not sure I follow multiple scripts part. Do you mean like semi colon separated or an explicit list of commands? I think if there is a good usecase I would take a look at it.

**Comment by MykolaGolubyev** [comment]
> Can you explain what the desired behaviour of `on` and `onOptional` are?

`on` will fail if there was no match during entire process run. `onOptional` (better name is welcome) is not going to fail if there was no match, but still lets you react (e.g. send `http.delete` to clear state).

File: webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java (hunk 4 above)
Comment (MykolaGolubyev) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: You were right, thanks @tsiq-karold. Wonder if executor would have been better in this case.

Comment (tsiq-karold) [reply webtau-cli/src/main/java/com/twosigma/webtau/cli/ProcessUtils.java]: That thought had occurred to me.  However, it won't prevent you from having to join conceptually, instead you'll call get on a future.  You'll also have to work out how to size the executor or I suppose let the JVM do that for you and hope it's never an issue (which for this usecase it likely won't be).

**Comment by tsiq-karold** [comment]
> > Do we want to support user supplied environment variables for the process? It's not uncommon for scripts to expect some vars. People may even wish to set a PATH so they can run all the tests without FQN for the scripts they're executing.
> > Do we want to support multiple script executions? Right now if you do two cli.run calls, they'll have totally separate environments (or shells or whatever you want to call it).
> > My hunch is yes to the env vars and no to the multiple scripts but curious on your thoughts.
> 
> Yes for env vars. Not sure I follow multiple scripts part. Do you mean like semi colon separated or an explicit list of commands? I think if there is a good usecase I would take a look at it.

Not sure I can think of a use case right now but yes conceptually semi colon separated or && separated is what I was thinking.  Generally that'd b the same as executing multiple cli.run sequentially anyway because unless you source the script it'll run in a subshell and therefore won't touch your env.

**Comment by tsiq-karold** [comment]
The env stuff looks good.  Did you want to tackle the args with spaces thing in this PR or create an issue and tackle later?  I'm happy to approve if we have an issue for it,

**Comment by MykolaGolubyev** [comment]
Let's do a separate PR: https://github.com/twosigma/webtau/issues/346

**Pull request merged**